### ISSUE TITLE
Feature Request: Copy the save file to the SD card without to open OSD 

### ISSUE THREAD

**Issue by caiosilva1993** [MiSTer-devel/Main_MiSTer#789]
Feature Request: Is it possible to implement autosave without the need to open the OSD menu to copy the save file to the SD card?

**Comment by sorgelig** [comment]
This is unsafe feature as it may write to SD card at any time, including the time when you may press reset or power MiSTer down.
This is why write is implemented in determined times (when you open OSD).

**Comment by machine-check-exception** [comment]
Hi
Just an idea: it could be interesting and option inside the osd, under the reboot option, for example, that says: turn off. That could do all the writes left (if any) and leave the system in a semi standby state. It would be great if it could auto shut off like pi4 (raspberry) does, for example, but that is probably a hardware limitation. Even if that button is almost placebo, it could help people to take the habit of open the osd and activate that option before shutting down the power (and hence not loose any nvram change)
Hope not to say any stupidity :)
Thanks

**Comment by sorgelig** [comment]
MISTer concept originally is about using as a non-computer device. So whenever you want it turn off then just turn off. Adding computer style menu option "Turn off" is opposite the main concept. By the way, there is no power control on the board, so it won't able to be turned off nor sleep programmatically. The only available option is to terminate all Linux processes and halt.

I found it funny you forget to just open OSD to automatically save the data but you want to have menu where you have to navigate to specific option ;)

**Comment by machine-check-exception** [comment]
Damn!! You missed my point!! My bad. Let me try again :D

You could create an option that says "turn off". That options **help** users to get **the habit** of: **open the OSD** (that's the important part, of course xD. I omitted that because I found it so obvious as it is the main purpose of the "turn off" option) and then select "turn off" (that can even be completely placebo; Just print a message saying "good bye" or, if needed/desired, do some work). 

I suspected about the hardware power control limitation :(

Jokes apart, I would like to add/link with your comment:

a) Agree with the primordial idea of being a stateless system but it looks that as far as we go, the systems recreated are getting more complex, for example with the nvram or maybe other techniques/technologies that are currently present or may be present in a future. It would be great that mister could behaves, for example, as a replacement for an arcade cabinet, but with the nvram stuff that most arcades have, you can't just turn on and at the end of the day turn it off without loosing data. If I'm correct, you loose at least the scores and if you do some changes in the service and don't open the OSD, those changes are also lost. So the stateless original idea, I think is "already lost".

b) there are systems like neogeo and psx that needs to open the OSD in order to be able to save the cards to the SD card if auto save is on, than if I remember well, in neogeo, by default, it is off. If users don't know what is all this about they will turn mister off and next time they turn it on, they will ask themselves what happened, why all configs, saves in the card and the last scores are lost. A "turn off" options could help to unify and trigger some saves that sometimes are not clear. Of course, probably, each core should be adapted to that new triggering hook, but hey, at least we would have solved the open OSD part :)

c) Rules are already set. We have what we got (nice hardware and lot of work in software) and we must be imaginative to overcome some limitations that eventually are appearing. My proposal is not a technical solution in this case, it takes a round and tries to minimize a problem with an (out of the box?) alternative solution: education (try to educate the user to do something)

All this was just an idea and I don't want to create more work on an already full of tasks workflow that some good samaritans are already doing for all of us. Thanks very much for your dedication! :)

**Comment by birdybro** [comment]
I think a Turn Off option in the OSD is a lot less intuitive overall. Just open the OSD and then you can change cores, or reset, or power off the MiSTer. Once the user learns about how autosave works, it's a non-issue.

**Comment by machine-check-exception** [comment]
Normally it is not an issue till you power off the mister without doing that last open gui action :(
As sorgelig said, writes without control (when they are done and how much often they are done) is not good. But in most cases we know when we want that memory to be saved: when changing options in the service, when storing scores and when saves on a card are made by the simulated system. Would be inadequate to try to identify those moments and trigger a save to microSD card when found? For example watch for a known nvram memory position and when something is written there, trigger a save to microSD. Those moments would coincide with going out of service, after introducing score name or when saving to card on the system. Don't know if an sporadic write to the microSD card can affect the performance of the simulation.

**Comment by sorgelig** [comment]
Stop saying "if user don't know". This is not an excuse! Every user has to learn the system they use. If you don't want to study the system to have a good experience then you will learn it with hard way by loosing your saves. This is normal for lazy people. 
MiSTer won't be a computer, so it's highly unlikely there will be a "turn off" option.
If you want a computer experience then you have such option with many emulators on PC. MiSTer is not a PC.

**Comment by machine-check-exception** [comment]
Allow me respond in sections:
a) _MiSTer won't be a computer, so it's highly unlikely there will be a "turn off" option._
Damn! I have the feeling I didn't get you to understand the point  :( Let me try with another example: why don't you add a new option under the reboot that says: "if you press here before power off you are gonna have 12 hours of luck" or, "press here before power off to get another lottery number" or "press here before power off and you will be rewarded with a bitcoin" (you pay ;)
Its about lure/incentivise people to do the action you want. You designed the system for saving those states in an proactive way (instead of an automatic way), then, following that path, try to create a small help, a step beyond, an user experience improve, for helping people to remember open the OSD before shutting down. The "power off" option had more logical sense than the examples exposed now, but the intention is not convert the mister in a computer with a "power off" option. But maybe doing so we get people to remember to open the OSD; power off, free bitcoins or whatever good idea that helps/lure people to remember to open the OSD. This way is stop thinking in technical solutions and start thinking in "social" solutions. And that is the suggestion, try to think social solutions for overcome technical limitations/decisions.

b) "_Every user has to learn the system they use_". I'm pretty sure, at some point, somebody said so and now we have two major OS systems and a minor OS system :(

c) "_This is normal for lazy people_". Assume or believe that bad things happens because people is lazy is very brave and in my opinion a mistake that denotates a narrow sight (as always no intention to offend). Not everybody is a prince without any obligation with the maximum dilemma of lazing around in a Ferrari or invest time learning something. People work many hours a day, and then go outside tired and then they have to go shopping or hunting for food, invest time in children answering silly questions over and over all the afternoon, then arrive home and clean the house, prepare dinner, clean themselves, maybe study something because they missed the opportunity when they were young and now they want to catch up... And in the middle of all this hell, people search for ways to escape from reality. Usually they take the easy path: social medias, streaming, reading books or playing whatever they can.  Others in the middle of that chaos find the mister project, see that thanks to new projects like mister pi or qmtech, this technology now is more affordable and watching the way technology prices are rising up nonstop it makes they think: "hey, maybe I can recreate part of my youth, with a good price vs entertainment and share it with my children". Then, at the end, they get a mister, and with some effort, they get it up and running (create the sd card and fill with roms is the major issue). They start playing maybe few minutes a week, maybe their children play, or maybe not. But somehow, something is odd. That save state stored in the psx card (psx core), sometimes is lost and sometimes not. And the same happens with other configs or saves. Its just a little frustrating, but not live or dead situation, and, because people used to have lotta other things to do, like in general survive to a shitty live that tries to screw you up in every corner, they just accept this frustrating fact and take it as another nail in the grave of live.

Luckily this is not my situation, but I'm not unaware that every user has it's own story and when (bad) things happen, it is not always due to laziness. People drive cars and they don't know mechanics, people uses TVs and they don't know electronics, people use computers and they are not computer engineers. They had to learn something, true, but the systems are designed to be user friendly, call it dummie-proof. It is not safe to expect that every user will have the time and ability to understand how everything works and it is not fair cause them pain because of this. I think we should help them, not punish them. Unless, of course, you hate them... is that the case :P? Do you prefer to keep mister ecosystem small for capable people or do you prefer to allow not so capable people to join the project as probably "dummy" users (with respect to everybody) and help them to have a (still more) satisfactory experience?

Of course all the above exposed is more like an allegory and extreme cases, trying to popup a different point of view over the users. All with irony and respect to everybody. My intention is not, by any means, to attack or offend: nor devs, nor users. Take it, please, as a funny relate.

**Comment by sorgelig** [comment]
On most cores LEDs are already indicating a pending save. Cannot tell for all cores as some of them i didn't touch, but cores i make releases they use LED to signal.

**Comment by machine-check-exception** [comment]
Now I'm addicted to take a look at the green light!! Damn!! (thanks for implementing this idea in advance). There are cores where there is no light and, when you open the osd, there is nothing about saving in progress. Is like you open the OSD because it was told to do so, but never sure if something happens except when you came back and the scores are there (or not). I must assume that every core is saving the nvram (at least hi scores?) to the SD (or any media in use) when the OSD is opened?

For those misters without leds or enclosed somewhere without visibility, would it be interesting or feasible the idea of an option for activating visual tips about the memory or other status?

When you created the architecture for interactions between mister core (it's a linux program running on the ARM, right?) and fpga cores, did you add any kind of api/routine for the cores to notify mister core about their status (pause, nvram, cards, test mode, etc)?

Thanks and sorry for the questions if they are answered anywhere else (the option of open the code and understand everything is coded there doesn't count as a place for seek for answers xD)

**Comment by sorgelig** [comment]
They way FPGA works is completely different than traditional CPU based system. Everything happens in parallel. No, there is no such thing like total information about core's internal state. Cores are very different from each other and have different states. Each core decides itself when and what to write, so it doesn't notify linux system unless that is actually pushed for writes.

**Comment by Sawtaytoes** [comment]
Things I'm seeing here as reaons to why this feature doesn't exist (saving when Backup RAM updates) don't make sense for my use case.

You note "it will kill the SD card", but can't we just bypass the SD card altogether?

1. What if I don't use an SD card and instead use something more resilient like an HDD or SSD?
2. Can I move my save files to an SMB share? While that'll add more latency, if I have a Ethernet, it should be negligible compared to the slow write speed of many SSDs. Write now, the CIFS script only mounts one share though, and I'm using mine for games, not saves. The other benefit of putting them on an SMB share is I can have 2 MiSTers that share saves. I also have a backup.
3. What if I use a security-camera SD card? Those are designed to have a ton of writes, so they'll be more resilient to failure. I've used them on Home Assistant without issue.

**Comment by birdybro** [comment]
You forgot the other reason. MiSTer doesn't have the capability of a safe-shutdown with regard to the integrity of the save data. If you turn it off and a save is occurring (which on some cores can be happening multiple times per second constantly), the data will be garbage. This isn't as much of an issue for software emulators because everything is kinda pre-buffered as far as states go and there's a lot more pre-processing that can be baked in to account for this. On the MiSTer it's more like an NES. You know how you can corrupt NES saves on a real cartridge on real hardware by just turning it off incorrectly? The hold reset before powering down is kind of the equivalent of having to open the OSD in this.

**Comment by Sawtaytoes** [comment]
What about this... Make it so you can put the device in standby. Instead of a physical power button, do a virtual one.

The issue is my NES does save at any time, so I can turn it off whenever because saves are instant. Same with GameBoy, etc. If the game says it's good, I'll know I'm good.

With that Remote extension or Zapadoo, my kids will scan an NFC card while I'm playing and kill hours of saves. This is also how Heber's Multisystem 2 will work.

Makes the MiSTer useless to me for games that have saves.

When switching Cores, it could first open the OSD and wait for a save before switching. But...

It's not a question of education. I _will_ mess this up because it's designed this way. Every console saves automatically to a storage medium. Even if you have to go save in the game, you don't also have to open the Xbox menu for it. It's not the same thing.

Best would be, like my previous post, to stop worrying about killing the SD Card and bypass it altogether. Save to battery-backed RAM or to another location where constant writes are safe. Then when the game says "don't turn off the console", that's all I have to worry about.

I'm not asking you change how it works today, I'm saying to allow having actual automatic Backup RAM saves as a feature.

**Comment by birdybro** [comment]
>With that Remote extension or Zapadoo, my kids will scan an NFC card while I'm playing and kill hours of saves. This is also how Heber's Multisystem 2 will work.

>Makes the MiSTer useless to me for games that have saves.

You are using an unofficial convenience script to load games via tapping an NFC enabled card and that somehow becomes the project's problem to solve? I don't have this problem because I don't have a TapTo/Zaparoo. How do you keep your kid from smacking the reset button on your NES before you save? How do you keep your kid from tapping the NES and it causing the game to hang due to pin misalignment temporarily? etc... You are describing a problem that is external to the project or it's design.

I agree that having true autosave without having to open the OSD would be a nice thing to have, but there are more reasons than what you listed.

Allowing it as a feature makes sense if you ignore that it will generate a whole bunch of people that then turn it on and complain to us about corrupted saves and microSD's that don't last multiple years anymore.

**Comment by Sawtaytoes** [comment]
I agree. That's troubling. It would suck to have users complain about enabling a simple setting and bricking their device. That's arguably worse than losing one game's progress.

It's why TrueNAS also stopped allowing you it install to USB drives and why Home Assistant changed their defaults. I understand the issue. 👍

Home Assistant let's you change those settings and even write to an NVMe drive (which I have). TrueNAS encourages you to install to an SSD mirror.

I don't think the setting needs to be easy to enable. It should be cumbersome because only people "in the know" should enable it. Then it'll be less likely to happen by a person who didn't know.

Is there a way I could add a script in there that does this for me? Maybe the setting could require a special file with a passcode to ensure it gets turned on? Or check the mounts and never allow writing to the SD Card?

My big issue is that this is **impossible to do today**. I simply want a way to enable it. Having used consoles and PCs for years, my intuition of "it says it saved" is very different to how the MiSTer does it.

I play games to de-stress and not think about things like that. If the game says it saved, I intuitively believe it saved. I don't have the time to replay the same 2-5 hours.

---

In terms of my kids turning thing off, I move power buttons in places they can't reach, so they don't turn things off. I'm deliberate about it.

Also, I need some sort of UI to load games because I want my kids to be able to play games on the MiSTer. That's why I bought a DE10 Nano and all these accessories. I wanna introduce them to retro gaming without them trashing my old consoles, original boxes, etc.

Last night, I was playing a bit of Link's Awakening and lost my save. I turned off the console without realizing it doesn't save unless you go to the OSD. I've already dealt with this, and it's insanely stressful to remember to do it when the game normally would have saved if it was a physical console.

**Comment by birdybro** [comment]
It's not as simple as writing a script to just do it for you AFAIK. You need Main to receive updated save data from the FPGA core over the FPGA to HPS interconnects at some kind of interval or in response to a signal, then you need main to update the save file with that.

**Comment by Sawtaytoes** [comment]
Darn. Then I'm back to asking that it be made an optional feature even if it's made difficult to enable.

**Comment by Sawtaytoes** [comment]
To workaround this issue for now, I set up something myself to open the OSD before switching games and before "turning off the console" by going back to the main menu and closing out the core.

I was wondering if you could tell me if this save happens async or if it waits to finish saving before swapping cores. I don't have enough info on how Autosave works. Is that documented somewhere?

**Comment by sorgelig** [comment]
If there is pending save, then upon opening OSD it will show "wait" screen. Once this screen is gone, it's safe to swap the core/game or turn it off.

**Comment by Tombert** [comment]
After losing about an hour of Phantasy Star IV because I forgot to open the OSD and then my power went out, I really think there should be a way to save without opening the menu. 

I think the concerns with SD card wear are really overblown but this could be ameliorated by putting it behind a configuration and mark it as "danger zone" or something if people want to accept the risk. 

I don't know as much about FPGAs as I do about software so forgive some ignorance on this, but is there any reason there couldn't be something that writes every N seconds, writes it to a temporary file, then do an atomic `mv` to use the temporary file to replace the current save file? At that point the concerns about "what happens if power fails in the middle of a write?" becomes a non-issue; if the power dies in the middle of the write, the original save file isn't affected until the write is done, so if a power outage happens mid-write it will only screw up the temporary file.  You'd lose up to N seconds of progress but it wouldn't be a as bad as people losing hours of saves because they forgot to hit the OSD. 

There's a few ways that the number of writes could be reduced if you're actually concerned about SD card wear; you could keep the "save on OSD open" feature to force a write for anything pending, but you could do something like hash the sram state and only write if it is divisible by 20 or something.  If there is enough state info to light up the LED for the pending sram to write, then I don't see why this couldn't be done. I say "hash" but it could be anything, just something that can work as a discriminator to not write too often.

**Comment by birdybro** [comment]
Please read the other comments above. It's not just about SD card wear solely.

**Comment by Tombert** [comment]
I did read the entire thread.  I personally do not find the other reasons terribly compelling. Additionally if you had read my entire comment you'd see that I addressed the concern of uncompleted writes in the event of a power failure. 

It's easy to say "YOU'RE USING A NEW THING SO YOU SHOULD BE WILLING TO LEARN HOW TO USE IT LOL I AM VERY SMART", but that's honestly a pretty dumb position to have.  The goal of the MiSTer is to faithfully run old games.  If I played Phantasy Star IV on an original Sega Genesis, I would save and than I could turn the console off.  I didn't have to then find some other special system menu every time I save the game. 

Sorg seems very willing to die on this hill for reasons that are unclear to basically everyone with reasoning that I don't actually think he even believes as he's writing it.

**Comment by birdybro** [comment]
Submit a pr

**Comment by Tombert** [comment]
I don't know how to write FPGA code.  I am sure I could learn it but I'm pretty confident that if I did submit a PR I'd get some sanctimonious comment repeating the same verbiage from above and then it would be rejected.

**Comment by birdybro** [comment]
You think this is the appropriate way to request features from open source contributors?

I have a concept of how it could be done in my head already, but it would require an update to the framework for the cores, then a custom update to every single core that uses saves, and any future core using saves would need to use the same method or the experience wouldn't be the same. And then Main_MiSTer would have to be updated to make it work. Because there isn't really a signal that can be solely read from Main_MiSTer's side to make it work. And then this would be a good time to update the developer documentation to show how saves are done since I don't remember if that is documented or not.

This method might cause problems in the PSX core though. If you are familiar with that core, it has to pause the core to update the save files for stability. It can be done without, but a few games might have issues without pausing. This is one exception, so what should we do then for this? For instance, there are some games that automatically update the memory card file, so should we pause the core every 10 seconds? Do we put something in the Main_MiSTer to say, if it's the psx core only pause the core every 1 minute? But would users really want the core to be paused every 60 seconds? That seems like a bad experience.

I'm genuinely curious how duckstation handles this alternately. Since it's running with massively larger memory buffer capability due to not being on FPGA it probably just holds the whole thing in a memory buffer and dumps periodically to a file if it's different or upon closing the program. If the program crashed or if your computer's power was shut off you might lose some progress.

The memory card file for the MiSTer PSX core is kinda like a shared object between the core's internal state for the memory, the linux filesystem, and any events (OSD save, mount, unmount, etc...). This can make it probably a little more than just having one signal and getting it to do a file update interval. What if during that update you are mounting a new drive in the OSD. What if someone has the pause when OSD open setting turned off etc...

How would you handle the problem of the data being updated by the core mid-update to linux fs?

I'm sure there's some satisfactory way to do it, but please don't act so entitled when you are ignorant about what goes into it. This is not a commercial product and we are not a customer service hotline.

**Comment by Tombert** [comment]
> I'm sure there's some satisfactory way to do it, but please don't act so entitled when you are ignorant about what goes into it. This is not a commercial product and we are not a customer service hotline.

I'm not acting entitled.  I didn't open the feature request, I was responding back to (what I perceived as) a passive aggressive "submit a pr" response you gave to me. Reading through this thread it feels like the responses have been a bit outwardly hostile when people make suggestions. 

I'm not new to the open source world, and I'm generally more polite about this stuff.  Please feel free to read through the thread again if you do not understand why I took the tone as hostile. 

Anyway, this is getting into high-school shouting match territory so please accept my apology if you weren't trying to be a jerk. 


--------

I didn't think about the Playstation core, so I don't know the difficulties with that.  I know that the SNES core doesn't stop the game when writing the SRAM so I didn't think that dumping that was a blocking operation, which is why I thought my "automatically write to a temporary file and atomically swap out the backup with the new one upon a successful write" thing could work.

**Comment by birdybro** [comment]
At the end of the day, if it was feasible to do in a safe satisfactory way for the user that wouldn't exceed the budget of resources on the FPGA side that current cores require, and that wouldn't generate more tickets, I trust someone would have done it by now, especially sorgelig himself. It's obviously a reasonable request.

The on paper "right" way to do this would be to have BRAM reserved for the save on the core side to act as a memory buffer that the core doesn't read to as it's internal state. Unfortunately a couple cores are at 100% bram utilization so that poses another limitation.

Sorry for my prickly response yesterday.

**Comment by Tombert** [comment]
Yeah, sorry for my pissy response as well. 

I'm not sure it's done the "right" way, but I saw someone with the SNES core to write a save every time the SRAM changes: https://github.com/Biduleman/SNES_MiSTer_DirectSave/

But I don't know how hacky that is.

**Comment by birdybro** [comment]
https://github.com/Biduleman/SNES_MiSTer_DirectSave/commit/61550d06d632a851f4a7c1fe4b297cb15d2f9a7f

This is risky, if there are transient state changes while the file is being written to the sd card, the save could potentially get corrupted.

**Comment by Tombert** [comment]
Yeah, that's why I thought it might be better to do it asynchronously; if the save status could be used to send a message/signal to something in linux-land, and then the write happens doing that atomic thing I mentioned in software land.  If you can get it done in software land then it's trivial to make sure that the write happens safely just using basic POSIX tools (as long as you're ok potentially losing some progress in the event of power loss or something). 

I realize it's probably not as simple as I'm describing. If saving the SRAM file is being handled in hardware-land then that's obviously harder.  I guess I thought since there's enough info to change the LEDs then this could conceivably be done fairly easily. I don't know the atomicity guarantees (if there are any) that can be had if the FPGA/core itself is the thing writing the files. I'm positive it's *possible* to do in FPGA-land but of course that could come at the expense of more synchronization code than there is budget for. 

Forgive ignorance on my end, but if you could indulge me, is it *possible* for the FPGA/core stuff to send some signal to the linux stuff? Because if it is then I think I could write the code to do the write safely.

**Comment by birdybro** [comment]
Couple draft PR's since I feel like I should try and explore this, looks simple enough in a way. This would make it identical to how bsnes does it by default pretty much, just a timed interval.

**Comment by birdybro** [comment]
Ultimately main problem here is primarily going to be CD-based cores that are using CD Audio, but the authors can choose not to implement it in their core (and probably shouldn't). You can see what would happen at every interval right now if you trigger a pending save in the MegaCD core and you can hear the CD Audio stuttering, like in Shining Force CD for instance. Because the File IO operations for the save congest the CD audio stream.

**Comment by Tombert** [comment]
That makes sense, redbook audio is expensive and so it makes sense that frequent requests could congest stuff. 

I think for SNES-and-earlier this would be less of an issue though?

**Comment by sorgelig** [comment]
If your area may have power outages, then the best way is to add UPS. As explained many times, not all systems have save storage as a memory card where same happened in a specific time. Many systems have storage as a battery equipped SRAM.

**Comment by NTKasper** [comment]
Hmm... is the only problem possibly corrupted files or is it about possibly corrupting a full SD card?

Because I would argue the problem with corrupted files could be solved with a temp file system. But that could get messy and depending on the core... well, it could end up taking a lot of space. I have wanted to see a feature like this but have so far not been able to come up with a good solution.

But pretty sure similar problems have been solved for other software many times. But if such a feature is implemented, it should certainly not do it often. Every minute might be too often and either way... it would be very useful if there could be some "stuff is being saved" type of icon in a corner maybe, like so many games have these days, with a warning to NOT turn off the system while that icon is showing. That could help with some of the problems.

I work with datawarehousing and sometimes to avoid such problems I do not write to a file that is used directly, I start with a temp table and then use it to overwrite the large file. The long process is writing the file itself, then renaming it to make it replace the old one is a matter of milliseconds... probably cannot even be measured in milliseconds.

With the MiSTer, maybe the whole file could start out being generated in RAM, then written to disk as a temp file, and then a rename to swap the existing file and the new file. That ought to minimize possible issues at least. But I suspect it would still be best as some opt-in features with a warning for people who really want this and knows the risks. Does not solve the possible problem of this then causing some lag or something for the active core...

**Comment by machine-check-exception** [comment]
Hi. In the past I proposed to overlay an icon on a corner of the screen
just reminding that there is data to be saved. With a visual tip it will be
more difficult to shut down the system without opening the osd. But in case
of power fail the problem will still be there...


El dt., 17 de febr. 2026, 10:44, Kasper Olesen ***@***.***>
va escriure:

**Comment by NTKasper** [comment]
Hi machine-check-exception,
Yeah, it would still be there, but my other proposal should help minimize that issue, by not making the new file directly but starting with a temp file, or even in memory, and then a temp file, and then renaming the files. It should help minimize the chances of dataloss or datacorruption. Not sure if you can end up with any special problems in the process of renaming files. But then it should in theory only be a problem with what name the file got? But hopefully it would either stop the renaming from happening or... maybe do a partial rename?

This is why I asked whether the bigger problem was potentially damaging the SD card or USB drive being written to because you cut power mid-write? But... well, if it is due to power loss that is a problem we could have either way. So not even sure how valid a concern that is. I guess it could be argued that automatic writes increases the chance by there being more writes. But I still think my solution would somewhat remove those issues.

Steps:
Write save to a temp file -> rename the current save file to a backup_temp file -> rename the new temp file to the name of the active save file

Should this process be broken, I guess a cleanup script might be possible to make that finds these broken processes and completes them. But I guess that temp save might be in risk of being corrupted.

**Comment by birdybro** [comment]
Let's just rule out sd wear entirely for the sake of argument, it's largely a non-issue as there is wear-leveling at the linux level for exFAT for microSD cards, and microSD cards are rated for far more writes than original flash memory like PSX memory cards were, etc... If all we cared about was microSD health and microSD sector corruption, exFAT is not the most optimal solution anyways, but it's a tradeoff due to it being mountable in windows.

The remaining potential issue is that your former good save gets replaced with a corrupted bad save and you have no way of even going back to the old good save, if you power off in the middle of an autosave. The only way to deal with this 

Regarding the temp file system as a power outage solution, exFAT doesn't support atomic writes natively (no journaling), so it would have to be some kind of db system for the save folder. It would either default to a good new save or a good old save, nothing in between. Determining what a "good save" is though, not sure how that can reliably be done. How this could be done is beyond me, way above my paygrade!

Regarding the overlay icon, in some cores this will constantly be on which is the problem. NES games that use save space as extra work RAM for instance.

**Comment by Tombert** [comment]
didn't realize the exfat doesn't have atomicity; I looked it up and it looks like you're correct; the POSIX `mv` command (even in exFAT) is atomic in that it's safe from multiple processes but it doesn't guarantee safety in a power crash.  Like everything by Microsoft it's built wrong 😄 . 

What if instead of replacing the save, we just do an "append only" sense.  Save files are something like "gamename_timestamp" and upon loading a game it loads the latest timestamp?  I

Then the periodic saves could still work, and that way, if the latest save is corrupted it can just load the previous one. Again, you might lose some progress but only N seconds of progress instead of everything if you forgot to hit the OSD. If you're afraid of too many save files piling up, there could be a garbage collection phase upon loading the game, where it trashes everything except the latest non-corrupt SRAM.  This GC step could be done upon successfully writing the SRAM to disk but I don't know if that could affect performance negatively. 


ETA: 

Regarding "what's a good save?", I think we could checksum with md5 or CRC the SRAM before writing it, and prepend that into the save file.  Then upon loading the save it's checked against the checksum, and if that fails then fall back to the previous one, and if there is no previous one assume that that there's no save file.  

The choice of hash doesn't really matter, you could literally just add up all the 1's in the binary and store an integer at the beginning and that would likely be good enough, though if there's budget for it I think CRC would probably be optimal. 

Obviously there are implementation details that aren't terribly important here; you could store the checksum in the filename  for example, since that might be easier to parse than actually modifying the binary that gets saved to the disk.

**Comment by birdybro** [comment]
I think you could avoid the timestamp naming method with a minimal db instead. The issue of IO congestion could exacerbate with this, so moving more into ddr3 until the action is final to be written would offset that.

**Comment by Tombert** [comment]
I don't know if I fully understand what "db" means here, but now I'm wondering if the simplest solution would be to use something like SQLite.  I'd need to some research to see their atomicity guarantees, but I think that they have special tricks to ensure crash-safety and whatnot.  

I think if we implemented my example, we'd largely be reinventing SQLite, because I realize what I described is basically just makeshift journaling that is probably performed better by a database with ACID like SQLite. 

I don't know how much of the SRAM dump is implemented in software, but if the writing part is mostly software I think that could work fine. 

ETA: 

I looked it up, and it looks like SQLite is public domain, so even better than regular FOSS with regards to licensing-not-being-an-issue.  I also saw that there are knobs you can turn that would allow it to be flush-heavy if that's needed to ensure safety, but I think even without those knobs it would still likely be in the "safe enough" category given that exfat itself appears to be inherently unsafe because metadata can be corrupted.

**Comment by Tombert** [comment]
https://github.com/MiSTer-devel/Main_MiSTer/pull/1099

Ok, so I got codex to hack together some basic POC stuff and it builds on my machine but I'm not 100% how to test this.  It compiles but I'm not sure the best way to flash this to an SD card to check to see if it works.

**Comment by Tombert** [comment]
Sorg doesn't like my solution, but for those of you who do not think that the current code is "perfect" like he does, you can look at my fork https://github.com/Tombert/Main_MiSTer. My fork uses SQLite under the covers, and it auto-writes the SRAM every 5 minutes by default, or immediately if you open the OSD. 

My PR was unceremoniously closed because it's not "perfect" but for those of you who want your devices to do what you expect them to do it's available for you.

**Comment by NTKasper** [comment]
There are a lot of things to take into account. One thing is dependency requirements and I suspect Sorg dislikes getting a SQLite dependency for the MiSTer. I think I will try your fork anyway though and also try to see what Sorg might dislike about it. If it does a lot of changes in several places that are difficult to follow, it is difficult to see the possible problems this might cause.

Seems Sorg closed this now? I will try to comment anyway

**Comment by NTKasper** [comment]
"Regarding the overlay icon, in some cores this will constantly be on which is the problem. NES games that use save space as extra work RAM for instance."

Sure, but I doubt the plan was ever to allow this to do constant saves. I guess it is a minor issue to remember to always open the menu before closing a core or turning off the mister, but still. But also the icon seems like a feature of its own in case maybe some save actually takes longer than a second or so when using the menu to save. Sure, the menu might technically lock up while it is saving, but if the user cannot see that, its hard to know when it is really done saving.

**Comment by Tombert** [comment]
@NTKasper I'm not married to SQLite, just my initial thing ended up reinventing a lot of filesystem and database stuff in order to work around exfat. I figured that SQLite has been around forever, is very low-level, has many tuning knobs, is *very* thoroughly tested, and would be easier to just get a POC. I am pretty sure I could get it working without that, but I wanted to do a basic smoke-test.  I didn't expect the SQLite to be merged, which is why I made it a draft PR so that my work could be open and then I would probably do it without SQLite. That said, SQLite seems to work fine, at least with the GBA, SNES, and Genesis.  I haven't thoroughly tested any disc based consoles yet, so there might be a performance degradation here.  


TO BE CLEAR, this is still heavy beta.  I take no responsibility if your saves get lost or corrupt.  Please back up your save files before attempting this. 

Frankly I have no further desire to contribute to anything Sorg runs.  I could elaborate why but I'm afraid it might involve some mean words that probably aren't appropriate to attach to my username.  If you read the comments on my PR, he seems to be under the (strange and incorrect) impression that all file IO is created equal, and he doesn't seem to understand atomicity at the filesystem level, which would be fine if he weren't so sanctimonious about it.

**Comment by Tombert** [comment]
@NTKasper  I'll see if I can add the save icon in my fork tomorrow.

**Comment by NTKasper** [comment]
I use SQLite in several projects and it certainly seems one of the best and pretty much only minimal databases. I was looking into alternatives last week I think it was. Its technically possible to make f.ex. Postgres portable but its just a whole other size to go with.

So if MiSTer is to have a database dependency it should be the one to go with I think.

**Comment by Tombert** [comment]
There are a few competitors now, like RocksDB and DuckDB I think.  I've never used DuckDB and I'm not sure how often it's used in embedded tasks so it didn't really come to mind.  I have used RocksDB a lot and RocksDB is cool and fast but it will eat all your memory. 

SQLite generally doesn't take much more overhead than writing files yourself, especially if you end up reinventing a lot of ACID to work around exfat's bullshit.

**Comment by Tombert** [comment]
Last post I will push here since I don't want to start many more flames than I already have.

I will be maintaining my own fork independent from the core MiSTer team as I do not agree with the governance of this project. 

My fork allows for automatic periodic saves as discussed, and additionally allows you to tag saves and restore them later, entirely within the UI. Additionally, it can import and export the current .sav files. The entirety of writes are append-only and checksummed, so versioning is automatic, and writing the SRAM is atomic meaning that a write either happens or it doesn't, which is likely what you want instead of corruption.  I am still in the process of testing this but I am getting very promising performance results and since I built the model correctly it's relatively easy to add features on top of it, which might be fun. 


I will be adding more features to improve write safety since Sorg seems completely unwilling to solve this problem and seems to think that "just buy a UPS!" is a solution to exFAT's lack of journaling.  Sorg is a smart guy but that's such a ridiculous statement that I have to think it's cognitive dissonance to justify not solving the problem, or because he doesn't actually realize that SQLite is an in-process library. The problem can be solved without SQLite, it could be solved without SQLite even that much effort actually, but Sorg has said the current "force the user to do it in the OSD is perfect", so he doesn't appear to want to solve these problems and apparently doesn't care if the current broken implementation corrupts your saves. 

Ugh, getting into flame territory again. Sorry. 

Anyway, if anyone actually wants the feature that this thread is about, or any of the features I mentioned above, check out my fork: https://github.com/Tombert/Main_MiSSus

Feel free to open feature requests and report bugs!
